pkg/security/security_profile/activity_tree/activity_tree.go

```diff
@@ -0,0 +1,381 @@
+		node.MatchedRules = model.AppendMatchedRule(node.MatchedRules, event.Rules)
+		return newProcessNode, nil
+	case model.FileOpenEventType:
+		return node.InsertFileEvent(&event.Open.File, event, generationType, at.Stats, at.shouldMergePaths, shadowInsertion)
+	case model.DNSEventType:
+		return node.InsertDNSEvent(event, generationType, at.Stats, at.DNSNames, shadowInsertion)
+	case model.BindEventType:
+		return node.InsertBindEvent(event, generationType, at.Stats, shadowInsertion)
+	case model.SyscallsEventType:
+		return node.InsertSyscalls(event, at.SyscallsMask)
+	}
+
+	return false, nil
+}
+
+// CreateProcessNode finds or a create a new process activity node in the activity dump if the entry
+// matches the activity dump selector.
+func (at *ActivityTree) CreateProcessNode(entry *model.ProcessCacheEntry, generationType NodeGenerationType, shadowInsertion bool) (node *ProcessNode, newProcessNode bool, err error) {
+	// drop processes with abnormal paths
+	if entry.GetPathResolutionError() != "" {
+		return nil, false, fmt.Errorf("process cache entry error: %s", entry.GetPathResolutionError())
+	}
+
+	// look for a ProcessActivityNode by process cookie
+	if entry.Cookie > 0 {
+		var found bool
+		node, found = at.CookieToProcessNode[entry.Cookie]
+		if found {
+			return node, false, nil
+		}
+	}
+
+	defer func() {
+		// if a node was found, and if the entry has a valid cookie, insert a cookie shortcut
+		if entry.Cookie > 0 && node != nil {
+			at.CookieToProcessNode[entry.Cookie] = node
+		}
+	}()
+
+	// find or create a ProcessActivityNode for the parent of the input ProcessCacheEntry. If the parent is a fork entry,
+	// jump immediately to the next ancestor.
+	parentNode, newProcessNode, err := at.CreateProcessNode(entry.GetNextAncestorBinary(), Snapshot, shadowInsertion)
+	if err != nil || newProcessNode && shadowInsertion {
+		return parentNode, newProcessNode, err
+	}
+
+	// if parentNode is nil, the parent of the current node is out of tree (either because the parent is null, or it
+	// doesn't match the dump tags).
+	if parentNode == nil {
+
+		// since the parent of the current entry wasn't inserted, we need to know if the current entry needs to be inserted.
+		if !at.validator.MatchesSelector(entry) {
+			return nil, false, nil
+		}
+
+		// go through the root nodes and check if one of them matches the input ProcessCacheEntry:
+		for _, root := range at.ProcessNodes {
+			if root.Matches(&entry.Process, at.differentiateArgs) {
+				return root, false, nil
+			}
+		}
+
+		// we're about to add a root process node, make sure this root node passes the root node sanitizer
+		if !at.isValidRootNode(&entry.ProcessContext) {
+			if !shadowInsertion {
+				at.Stats.droppedCount[model.ExecEventType][invalidRootNodeReason].Inc()
+			}
+			return nil, false, fmt.Errorf("invalid root node")
+		}
+
+		// if it doesn't, create a new ProcessActivityNode for the input ProcessCacheEntry
+		node = NewProcessNode(entry, generationType)
```

Comment:
same remark as before, IMO we should not call NewProcessNode if !shadowInsertion